### PR TITLE
PP-742: Scheduler crashes during re-confirmation..

### PR DESCRIPTION
**Pull request by agrawalravi90**
#### Issue-ID

* **[PP-742](https://pbspro.atlassian.net/browse/PP-742)**

#### Problem description
* Scheduler crashes during re-confirmation of a degraded standing reservation

#### Cause / Analysis
* PP-84 introduced a bug inside query_reservations(), resresv_arr[] was not being created correctly.

#### Solution description
* Fixed the code so now resresv_arr[] gets created correctly

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_reservations.py [hunk 1]
```diff
@@ -42,6 +42,226 @@
...
+        r2_node = self.server.reservations[rid2].get_vnodes()[0]
+        self.server.expect(RESV, {'resv_nodes': (MATCH_RE, re.escape(other))},
+                           id=rid2)
+
+        if _m == PTL_API:
+            self.server.set_op_mode(PTL_API)
+
+    def test_degraded_standing_reservations(self):
+        """
+        Verify that degraded standing reservations are reconfirmed on
```
Comment (mike0042) [inline test/tests/functional/pbs_reservations.py]: There is a trailing space on this line causing the Travis checks to fail.

**Comment by mike0042** [comment]
We may need help from @subhasisb to resolve the license/cla check.

**Comment by subhasisb** [comment]
Ya, not sure where the second cla check is coming from. I will check it asap.

**Comment by agrawalravi90** [comment]
@bhroam requesting you to review these changes,  thanks!

File: test/tests/functional/pbs_reservations.py [hunk 2]
```diff
@@ -42,6 +42,226 @@
...
+        rid = self.server.submit(r)
+        try:
+            a = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+            d = self.server.expect(RESV, a, rid)
+        except PtlExpectError, e:
+            d = e.rv
+        return d
+
+    def test_no_span_psets(self):
+        self.create_nodes()
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: All the tests of this test suite don't have doc strings.

File: test/tests/functional/pbs_reservations.py [hunk 3]
```diff
@@ -42,6 +42,226 @@
...
+        a = {'do_not_span_psets': '1'}
+        self.server.manager(MGR_CMD_SET, SCHED, a, "")
+        d = self.submit_resv()
+        self.assertTrue(d)
+
+    def test_over_limit(self):
+        self.create_nodes()
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: I don't understand the purpose of this test.  Is it testing the fact that user limits do not affect reservation confirmation?

File: test/tests/functional/pbs_reservations.py [hunk 4]
```diff
@@ -42,6 +42,226 @@
...
+        a = {'max_run_res.ncpus': '[u:' + str(TEST_USER) + '=1]'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+        d = self.submit_resv()
+        self.assertTrue(d)
+
+    def test_resv_and_job(self):
+        self.create_nodes()
+        a = {'max_run_res.ncpus': '[u:' + str(TEST_USER) + '=3]'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: What's the purpose of setting this limit?  The test only submits a 1 cpu job.

File: test/tests/functional/pbs_reservations.py [hunk 5]
```diff
@@ -42,6 +42,226 @@
...
+    def test_round_robin_reservation(self):
+        a = {'resources_available.ncpus': 4}
+        self.server.create_vnodes('vnode', a, 3, self.mom,
+                                  usenatvnode=True)
+
+        self.scheduler.sched_config[
+            'node_sort_key'] = "\"ncpus HIGH unused\" ALL"
+        self.scheduler.apply_config()
+        r1 = Reservation(TEST_USER)
+        r1.set_attributes({'reserve_start': int(time.time() + 61)})
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: doesn't this need a duration or a reserve_end?  Is there a default it is using?

File: test/tests/functional/pbs_reservations.py [hunk 6]
```diff
@@ -42,6 +42,226 @@
...
+        r1id = self.server.submit(r1)
+        r1id = r1id.split()[0]
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: I don't understand this split().  The rid that will come back from submit() will be like R123.server.  I could understand split('.') maybe, but split() by itself will split on whitespace.  There shouldn't be any.

File: test/tests/functional/pbs_reservations.py [hunk 7]
```diff
@@ -42,6 +42,226 @@
...
+        a = {'job_state': 'R'}
+        self.server.expect(JOB, a, jid)
+
+    def test_round_robin_reservation(self):
+        a = {'resources_available.ncpus': 4}
+        self.server.create_vnodes('vnode', a, 3, self.mom,
+                                  usenatvnode=True)
+
+        self.scheduler.sched_config[
+            'node_sort_key'] = "\"ncpus HIGH unused\" ALL"
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: Why bother setting node_sort_key if you're not checking the order that nodes in the resv_nodes.

File: test/tests/functional/pbs_reservations.py [hunk 8]
```diff
@@ -42,6 +42,226 @@
...
+        elif 'PBS_TZID' in os.environ:
+            tzone = os.environ['PBS_TZID']
+        else:
+            self.logger.info('Missing timezone, using America/Los_Angeles')
+            tzone = 'America/Los_Angeles'
+
+        a = {'Resource_List.select': select,
+             ATTR_resv_rrule: rrule,
+             ATTR_resv_timezone: tzone,
+             ATTR_resv_standing: '1',
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: Is this required?  Won't the fact that there is an rrule be enough to say this is a standing reservation?

File: test/tests/functional/pbs_reservations.py [hunk 9]
```diff
@@ -42,6 +42,226 @@
...
+
+        self.server.status(RESV)
+        r1_node = self.server.reservations[rid1].get_vnodes()[0]
+        r2_node = self.server.reservations[rid2].get_vnodes()[0]
+
+        self.assertEqual(r1_node, r2_node)
+
+        a = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        self.server.expect(
+            RESV, a, id=rid2, offset=36, interval=1, max_attempts=20)
```
Comment (bhroam) [inline test/tests/functional/pbs_reservations.py]: Wasn't this already done at line 193?

**Comment by mike0042** [comment]
@agrawalravi90: Please go ahead and squash your commits. The only change in the second commit was removal of a space to make Travis work. @bhroam can then merge once he signs off.

**Comment by agrawalravi90** [comment]
Hey Bhroam, thanks for being particular about the tests, but I couldn't really answer your questions as I had not written them myself, so, in the interest of time, I've removed all of them except the one which is relevant to this bug, please let me know if it looks okay.

**Comment by bhroam** [comment]
Looks good to me.  I'll merge it in.

**Pull request merged**